### PR TITLE
Common: Replace ip2nation by ip2location.

### PR DESCRIPTION
**Pull request by Gooyeth**
**Changes proposed:**

-  Replace ip2nation due to outdated.
- Now IP2Location is used
- Support IPV4 CSV **only**. [[Download]](https://download.ip2location.com/lite/)

**Target branch(es):** 3.3.5/master

**Tests performed:** Build and tested ingame

### CODE REVIEW

**Comment by Shauren** [comment]
but, why?

File: src/common/IPLocation/IPLocation.cpp [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "IPLocation.h"
+#include "Config.h"
+#include "ByteConverter.h"
+#include "IpAddress.h"
+#include "Log.h"
+
+IPLocation::IPLocation() : _fileHandle(nullptr), _databaseType(0), _databaseColumn(0), _databaseYear(0),
```
Comment (illfated) [inline src/common/IPLocation/IPLocation.cpp]:
```cpp
/src/common/IPLocation/IPLocation.cpp:24:88:
 fatal error: field '_databaseYear' will be initialized after field '_databaseMonth' [-Wreorder]
IPLocation::IPLocation() : _fileHandle(nullptr), _databaseType(0), _databaseColumn(0), _databaseYear(0),
                                                                                       ^
1 error generated.
```

**Comment by Shauren** [comment]
use the csv format

**Comment by Gooyeth** [comment]
@Shauren: Ready, thanks.

File: src/common/IPLocation/IPLocation.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+struct Iplocation
+{
+    uint64 ip_from;
+    uint64 ip_to;
+    std::string country_code;
+    std::string country_name;
+};
+
+class TC_COMMON_API IPLocation
```
Comment (joschiwald) [inline src/common/IPLocation/IPLocation.h]: having a struct `Iplocation` and a class `IPLocation` is a bit confusing, it would prefer `IpLocationRecord` and `IpLocationStore`

File: src/common/IPLocation/IPLocation.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        ip_to.erase(std::remove(ip_to.begin(), ip_to.end(), '"'), ip_to.end());
+        country_code.erase(std::remove(country_code.begin(), country_code.end(), '"'), country_code.end());
+        country_name.erase(std::remove(country_name.begin(), country_name.end(), '"'), country_name.end());
+
+        // Convert country code to lowercase
+        std::transform(country_code.begin(), country_code.end(), country_code.begin(), ::tolower);
+
+        Iplocation data;
+        data.ip_from        = (uint64)atoll(ip_from.c_str());;
+        data.ip_to          = (uint64)atoll(ip_to.c_str());;
```
Comment (joschiwald) [inline src/common/IPLocation/IPLocation.cpp]: `atoull`

File: src/common/Asio/IpAddress.h [hunk 4]
```diff
@@ -28,7 +28,7 @@ namespace Trinity
 #if BOOST_VERSION >= 106600
         using boost::asio::ip::make_address;
         using boost::asio::ip::make_address_v4;
-        inline uint32 address_to_uint(boost::asio::ip::address_v4 const& address) { return address.to_uint(); }
+        inline uint64 address_to_uint(boost::asio::ip::address_v4 const& address) { return address.to_uint(); }
```
Comment (Shauren) [inline src/common/Asio/IpAddress.h]: why was this changed? ipv4 address always fits into uint32

Comment (Gooyeth) [reply src/common/Asio/IpAddress.h]: Based on the csv file, exceeds the value uint32:

```
3758096128,"3758096383","AU","Australia"
```

Comment (Shauren) [reply src/common/Asio/IpAddress.h]: No, its impossible, ipv4 was designed to use exactly 4 bytes
also, 3758096128 < 4294967295 (it fits)

File: src/common/IPLocation/IPLocation.h (hunk 2 above)
Comment (Gooyeth) [reply src/common/IPLocation/IPLocation.h]: `IpLocationStore` for the class?

Comment (joschiwald) [reply src/common/IPLocation/IPLocation.h]: yes

File: src/common/IPLocation/IPLocation.cpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    TC_LOG_INFO("server.loading", ">> Loaded %u entries.", uint32(_ipLocationStore.size()));
+}
+
+Iplocation* IPLocation::GetData(std::string const& ipAddress)
+{
+    if (_ipLocationStore.empty())
+        return nullptr;
+
+    uint64 ip = Trinity::Net::address_to_uint(Trinity::Net::make_address_v4(ipAddress));
+    for (auto itr = _ipLocationStore.begin(); itr != _ipLocationStore.end(); ++itr)
```
[the remark this answers is not in the document]
Comment (Gooyeth) [reply src/common/IPLocation/IPLocation.cpp]: @Shauren: You can download the list to see it here https://download.ip2location.com/lite/IP2LOCATION-LITE-DB1.CSV.ZIP

File: src/server/worldserver/worldserver.conf.dist [hunk 6]
```diff
@@ -208,15 +208,6 @@ SourceDirectory  = ""
...
 
-#
-#    IPLocationFile
-#        Description: The path to your IP2Location database CSV file.
-#        Example:     "C:/Trinity/IP2Location.csv"
-#                     "/home/trinity/IP2Location.csv"
-#        Default:     "" - (Disabled)
-
-IPLocationFile = ""
-
```
[the remark this answers is not in the document]
Comment (Gooyeth) [reply src/server/worldserver/worldserver.conf.dist]: Simplifly the implementation in auth and world.

Suggestion of @Aokromes.

[the remark this answers is not in the document]
Comment (joschiwald) [reply src/server/worldserver/worldserver.conf.dist]: the location of external files should be configurable like all other files

[the remark this answers is not in the document]
Comment (Aokromes) [reply src/server/worldserver/worldserver.conf.dist]: tdb location is hard located there is no problem of continue like that on this even less when it needs to edit 2 files to set the location of one file.

[the remark this answers is not in the document]
Comment (Aokromes) [reply src/server/worldserver/worldserver.conf.dist]: the file size is really small, a lot of sql files are far larger than this, also, if you complain about file size we can turn to bin file :P

-rwxr-xr-x 1 aokromes aokromes 1.7M Apr  1 14:00 IP2Location.bin
-rwxr-xr-x 1 aokromes aokromes 6.6M Jun  1 14:00 IP2LOCATION.CSV

File: src/server/authserver/authserver.conf.dist [hunk 7]
```diff
@@ -157,6 +157,15 @@ SourceDirectory  = ""
...
 MySQLExecutable = ""
 
+#
+#    IPLocationFile
+#        Description: The path to your IP2Location database CSV file.
+#        Example:     "C:/Trinity/IP2Location.csv"
+#                     "/home/trinity/IP2Location.csv"
+#        Default:     "" - (Disabled)
+
+IPLocationFile = "./IP2LOCATION-LITE-DB1.CSV"
```
Comment (joschiwald) [inline src/server/authserver/authserver.conf.dist]: one suggestion: change full file location to folder only

File: src/server/authserver/authserver.conf.dist [hunk 8]
```diff
@@ -157,6 +157,15 @@ SourceDirectory  = ""
 
 MySQLExecutable = ""
 
+#
+#    IPLocationFile
+#        Description: The path to your IP2Location database CSV file.
+#        Example:     "C:/Trinity/IP2Location.csv"
+#                     "/home/trinity/IP2Location.csv"
+#        Default:     "" - (Disabled)
```
Comment (illfated) [inline src/server/authserver/authserver.conf.dist]: Won't the default be "current directory" or "current core directory" here?
```
#        Default:     IPLocationFile = "." (current core directory)
```
(this is just my suggestion, the format may be slightly inaccurate)

**Review by ghost — APPROVED** [review]

**Comment by Aokromes** [comment]
Any idea of why this?

) (Locked: No)xxx.xxx.xxx (Country

It fails like that on linux console.

Ingame:
¦ Last IP: xxx.xxx.xxx.xxx (Country ) (Locked: No)

(notice space after country name)

**Comment by illfated** [comment]
`) (Locked: No)xxx.xxx.xxx (Country`
That looks like output reached end of line, but restarted the same line (missing NewLine/Line Feed). I don't know if the code reads differently for in-game output compared to worldserver console.
Reminds me of some kind of invalid character set issue.

**Comment by Gooyeth** [comment]
@Aokromes, test https://github.com/TrinityCore/TrinityCore/pull/21957/commits/0e66f017654fabfe5dd07ed7182bbae623bec089, please
